src/core/file.cc

```diff
@@ -383,8 +383,14 @@ posix_file_impl::list_directory(std::function<future<> (directory_entry de)> nex
 
     // From getdents(2):
     struct linux_dirent64 {
-        ino64_t        d_ino;    /* 64-bit inode number */
-        off64_t        d_off;    /* 64-bit offset to next structure */
+        /* 64-bit inode number */
+#ifdef __USE_FILE_OFFSET64
+        ino64_t        d_ino; /* 64-bit inode number */
+	off64_t        d_off; /* 64-bit offset to next structure */
+#else
+        ino_t          d_ino; /* 64-bit inode number */
+        off_t          d_off; /* 64-bit offset to next structure */
```

Comment:
i think musl libc has not dropped the legacy types for large file support. see https://git.musl-libc.org/cgit/musl/tree/include/dirent.h#n59 . they are not provided by `_GNU_SOURCE` feature macro anymore, but under `_LARGEFILE64_SOURCE`. but yeah, they will be eventually removed. see https://musl.libc.org/releases.html for the release note for v1.2.4.

so, in the light of that `ino64_t` and `off64_t` are glibc extensions, and they are defined when `__USE_LARGEFILE64` is defined. and glibc also has type aliases of `ino_t` and `off_t` for these types if `_FILE_OFFSET_BITS` is 64. this macro is set by user.

`__USE_FILE_OFFSET64` is used by the glibc internally, it is defined when `_FILE_OFFSET_BITS` is 64.  if `__USE_FILE_OFFSET64` is defined, glibc defines 64bit APIs as default. which means the types like `ino_t` would be 64-bit, and `readdir()` would be 64-bit as well. one don't have to use the `readdir64()` variant explicitly.

so,

1. we should avoid using a macro used by glibc internally as a part of its public interface.
2. and the guarded statements are not quite relevant. as we should be able to use `ino_t` in the "64-bit" mode. as we want to use 64-bit inode number and offset.

as IMHO, the 64-bit inode number and 64-bit offset should be the default in this context as Seastar is designed to be used to build high-performance services, and they should be default under most circumstances. the existence their 32-bit cousins is only for the 32-bit systems -- only for embedded and some customized systems.

so, i think it'd be better to use `ino_t` and `off_t` directly. to achieve this goal, we can just put
```c++
static_assert(sizeof(ino_t) == 8, "large file support not enabled");
static_assert(sizeof(off_t) == 8, "large file support not enabled");
```

as by default, if we compile the tree with gcc or clang, `_GNU_SOURCE` is always defined.  and libstdc++ actually needs this, see https://gcc.gnu.org/onlinedocs/libstdc++/faq.html#faq.predefined. and glibc would define `_LARGEFILE64_SOURCE` if this macro is defined. and musl libc just use the 64-bit `ino_t` and `off_t`. so we should be safe in both cases without defining `_FILE_OFFSET_BITS=64` explicitly.

Reply:
okay will chenage